### PR TITLE
samples: matter: Added weather station sample

### PR DESCRIPTION
**Pull request by kkasperczyk-no**
Added weather station sample for Thingy:53 that enables creating
Matter device sending periodically temperature, relative humidity
and pressure measurements.

Signed-off-by: Kamil Kasperczyk <kamil.kasperczyk@nordicsemi.no>

### CODE REVIEW

File: samples/matter/weather_station/src/app_task.cpp [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+
+#include <app/common/gen/attribute-id.h>
+#include <app/common/gen/attribute-type.h>
+#include <app/common/gen/cluster-id.h>
+#include <app/server/OnboardingCodesUtil.h>
+#include <app/server/Server.h>
+#include <app/util/attribute-storage.h>
+
+#include <drivers/sensor.h>
+
```
Comment (Damian-Nordic) [inline samples/matter/weather_station/src/app_task.cpp:19]: pedantic: the empty line is unnecessary here (all headers are from the NCS).

File: samples/matter/weather_station/src/chip_project_config.h [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+ *          to the default configuration values for general CHIP features.
+ *
+ */
+
+#pragma once
+
+/**
+ * CHIP_DEVICE_CONFIG_DEVICE_PRODUCT_ID
+ *
+ * 0x4E4B: nrfconnect lock
```
Comment (Damian-Nordic) [inline samples/matter/weather_station/src/chip_project_config.h:21]: it's wrong, should be 4E57: nrfconnect weather-station

**Review by Damian-Nordic — APPROVED** [review]

File: samples/matter/weather_station/src/chip_project_config.h (hunk 2 above)
Comment (kkasperczyk-no) [reply samples/matter/weather_station/src/chip_project_config.h]: Oh right, copy paste :)

**Review by PerMac — CHANGES_REQUESTED** [review]
I don't see a sample.yaml in your sample. This file is required for regression testing based on Twister. This will allow the CI to test a given sample (at least by building it). Please add sample.yaml with a proper description to this sample. Let me know if you'd need a help with it.

**Comment by kkasperczyk-no** [comment]
> I don't see a sample.yaml in your sample. This file is required for regression testing based on Twister. This will allow the CI to test a given sample (at least by building it). Please add sample.yaml with a proper description to this sample. Let me know if you'd need a help with it.

@PerMac thanks, could you check if it's ok now?

**Comment by PerMac** [comment]
Looks good. I removed my +/-.

**Review by PerMac — APPROVED** [review]

**Comment by greg-fer** [comment]
Adding DNM and doc-required. Can we clarify why this sample does not have the docs required? I cannot allow a sample to go in without docs.

**Comment by greg-fer** [comment]
While we wait for the outcome of the discussion about the sample docs, I've added a bare-bones readme and created the follow-up JIRA: https://projecttools.nordicsemi.no/jira/browse/KRKNWK-10789

**Comment by kkasperczyk-no** [comment]
Thanks @greg-fer for contributing docs, I squashed it into my commit.

File: samples/matter/weather_station/README.rst [hunk 3]
```diff
@@ -0,0 +1,154 @@
+.. _matter_weather_station_sample:
+
+Matter: Weather station
+#######################
+
+.. contents::
+   :local:
+   :depth: 2
+
+This weather station sample demonstrates the usage of the :ref:`Matter <ug_matter>` (formerly Project Connected Home over IP, Project CHIP) application layer to build a weather station device that lets you remotely gather different kind of data, such as like temperature, air pressure, and relative humidity, using the device sensors.
```
Comment (Damian-Nordic) [inline samples/matter/weather_station/README.rst:10]: "such as like" ?

File: samples/matter/weather_station/README.rst [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+The sample supports the following development kits:
+
+.. table-from-rows:: /includes/sample_board_rows.txt
+   :header: heading
+   :rows: thingy53_nrf5340
+
+To program and gather logs from the Thingy:53 device, you need the external J-Link programmer.
+If you own an nRF5340 DK that has an on-board J-Link programmer, you can also use it for this purpose.
+
+For remote testing scenarios, if you want to commission the lock device and :ref:`control it remotely <matter_lock_sample_network_mode>` through a Thread network, you also need the following:
```
Comment (Damian-Nordic) [inline samples/matter/weather_station/README.rst:26]: "lock" -> "weather station"

File: samples/matter/weather_station/README.rst [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+
+* A Matter controller device :ref:`configured on PC or smartphone <ug_matter_configuring>`.
+
+.. note::
+    |matter_gn_required_note|
+
+Overview
+********
+
+The sample uses buttons for changing the lock and device states, and LEDs to show the state of these changes.
```
Comment (Damian-Nordic) [inline samples/matter/weather_station/README.rst:36]: There's just one button on Thingy:53. It's used for opening pairing window (if pressed for < 3s) or the factory reset (if pressed for >6s).

File: samples/matter/weather_station/README.rst [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+Remote testing in a network
+===========================
+
+.. include:: ../lock/README.rst
+    :start-after: matter_door_lock_sample_remote_testing_start
+    :end-before: matter_door_lock_sample_remote_testing_end
+
+.. _matter_weather_station_test_mode:
+
+Remote testing using test mode
```
Comment (Damian-Nordic) [inline samples/matter/weather_station/README.rst:56]: There's no the test mode for this example given we have just one button available.

File: samples/matter/weather_station/README.rst [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+.. _matter_weather_station_sample_remote_control_commissioning:
+
+Commissioning the device
+------------------------
+
+.. include:: ../lock/README.rst
+    :start-after: matter_door_lock_sample_commissioning_start
+    :end-before: matter_door_lock_sample_commissioning_end
+
+Upgrading the device firmware
```
Comment (Damian-Nordic) [inline samples/matter/weather_station/README.rst:133]: DFU is also unsupported for this sample.

File: samples/matter/weather_station/README.rst [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+You can use this sample as a reference for creating your own application.
+
+Requirements
+************
+
+The sample supports the following development kits:
+
+.. table-from-rows:: /includes/sample_board_rows.txt
+   :header: heading
+   :rows: thingy53_nrf5340
```
Comment (kkasperczyk-no) [inline samples/matter/weather_station/README.rst:21]: It should be `thingy53_nrf5340_cpuapp`

File: samples/matter/weather_station/README.rst [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+.. include:: ../lock/README.rst
+    :start-after: matter_door_lock_sample_test_mode_start
+    :end-before: matter_door_lock_sample_test_mode_end
+
+Configuration
+*************
+
+|config|
+
+Device Firmware Upgrade support
```
Comment (kkasperczyk-no) [inline samples/matter/weather_station/README.rst]: Thingy:53 doesn't support DFU yet.

File: samples/matter/weather_station/README.rst [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+===============================
+
+.. include:: ../lock/README.rst
+    :start-after: matter_door_lock_sample_build_with_dfu_start
+    :end-before: matter_door_lock_sample_build_with_dfu_end
+
+User interface
+**************
+
+.. include:: ../lock/README.rst
```
Comment (kkasperczyk-no) [inline samples/matter/weather_station/README.rst:56]: I would remove it and add instead something like that:

> 
Button:
    Depending on how long you press the button:

    * If pressed for 6 seconds, it initiates the factory reset of the device.
      Releasing the button within the 6-second window cancels the factory reset procedure.
    * If pressed for less than 3 seconds, it starts the the NFC tag emulation, enables Bluetooth LE advertising for the predefined period of time, and makes the device discoverable over Bluetooth LE.This button is used during the :ref:`commissioning procedure <matter_lock_sample_remote_control_commissioning>`.

File: samples/matter/weather_station/README.rst [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+    Optionally used for obtaining the commissioning information from the Matter accessory device to start the :ref:`commissioning procedure <matter_weather_station_sample_remote_control_commissioning>`.
+
+Building and running
+********************
+
+.. |sample path| replace:: :file:`samples/matter/weather_station`
+
+.. include:: /includes/build_and_run.txt
+
+See `Configuration`_ for information about building the sample with the DFU support.
```
Comment (kkasperczyk-no) [inline samples/matter/weather_station/README.rst:73]: To remove

File: samples/matter/weather_station/README.rst [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+    :start-after: matter_door_lock_sample_remote_control_start
+    :end-before: matter_door_lock_sample_remote_control_end
+
+.. _matter_weather_station_sample_remote_control_commissioning:
+
+Commissioning the device
+------------------------
+
+.. include:: ../lock/README.rst
+    :start-after: matter_door_lock_sample_commissioning_start
```
Comment (kkasperczyk-no) [inline samples/matter/weather_station/README.rst:98]: Generally that section should work fine for Thingy with one difference: ` ... a QR code, printed to the UART console ...` - Thingy:53 doesn't have UART console. It uses RTT interface.

File: samples/matter/weather_station/README.rst [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+.. |sample path| replace:: :file:`samples/matter/weather_station`
+
+.. include:: /includes/build_and_run.txt
+
+See `Configuration`_ for information about building the sample with the DFU support.
+
+Testing
+=======
+
+You can either test the sample's :ref:`basic features <matter_light_bulb_sample_basic_features_tests>` or use the light switch sample to test the light bulb's :ref:`communication with another device <matter_light_bulb_sample_light_switch_tests>`.
```
Comment (kkasperczyk-no) [inline samples/matter/weather_station/README.rst]: As discussed F2F for now I would mention that currently sample is still in development and cannot be tested, because we are lacking of proper mobile app with data visualization.

File: samples/matter/weather_station/README.rst [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+The sample supports the following development kits:
+
+.. table-from-rows:: /includes/sample_board_rows.txt
+   :header: heading
+   :rows: thingy53_nrf5340
+
+To program and gather logs from the Thingy:53 device, you need the external J-Link programmer.
+If you own an nRF5340 DK that has an on-board J-Link programmer, you can also use it for this purpose.
+
+For remote testing scenarios, if you want to commission the weather station device and :ref:`control it remotely <matter_weather_station_sample_remote_control>` through a Thread network, you also need the following:
```
Comment (kkasperczyk-no) [inline samples/matter/weather_station/README.rst:26]: In believe that here we don't have "no remote scenario", so I would rephrase it as mandatory requirement.

File: samples/matter/weather_station/README.rst [hunk 15]
```diff
@@ -0,0 +1,117 @@
...
+
+* A Matter controller device :ref:`configured on PC or smartphone <ug_matter_configuring>`.
+
+.. note::
+    |matter_gn_required_note|
+
+Overview
+********
+
+The sample can be tested in the following ways:
```
Comment (kkasperczyk-no) [inline samples/matter/weather_station/README.rst:36]: How about:

> The sample uses single button for controlling the device state.
It can be tested remotely over the Thread protocol, which requires additional Matter controller device that can be configured either on PC or mobile (for remote testing in a network).
The weather station device is periodically performing temperature, air pressure and relative humidity measurements, which results can be read using the Matter controller.
It communicates with the weather station device over Matter protocol utilizing Zigbee Cluster Library in order to describe data measurements within the proper clusters. 
To enable testing you need to complete :ref:`building and running the sample <matter_lock_sample_remote_control>`.
?

**Review by Damian-Nordic — APPROVED** [review]

**Review by greg-fer — APPROVED** [review]

File: samples/matter/weather_station/weather-station.zap [hunk 16]
```diff
@@ -0,0 +1,6998 @@
+{
```
Comment (rlubos) [inline samples/matter/weather_station/weather-station.zap:1]: What is this `zap` file, other samples don't seem to include it?

Comment (kkasperczyk-no) [reply samples/matter/weather_station/weather-station.zap:1]: It's a file that is used by the ZAP tool to generate files located in the `gen` directory. Other samples doesn't have it, as they have their corresponding light and lock samples in the upstream, so .zap files are located in there and that one sample is unique for the NCS. Nevertheless that was mistake from our side that we didn't include those .zap files for all samples from beginning and we are going to fix that soon, so all NCS samples will contain such files to allow easily regenerating `gen` directory content.

Comment (rlubos) [reply samples/matter/weather_station/weather-station.zap:1]: Ok, thanks for exaplaination. Does this format support some kind of comments? Would it be possible to add license header to it?

Comment (kkasperczyk-no) [reply samples/matter/weather_station/weather-station.zap:1]: I'm afraid not, this seems to be format inherited from .json that also doesn't support comments I guess.

Comment (rlubos) [reply samples/matter/weather_station/weather-station.zap:1]: ok, thanks for confirming 👍

**Review by rlubos — APPROVED** [review]
